proof_ipa.go

```diff
@@ -79,36 +82,68 @@ func GetCommitmentsForMultiproof(root VerkleNode, keys [][]byte, resolver NodeRe
 	return root.GetProofItems(keylist(keys), resolver)
 }
 
-func MakeVerkleMultiProof(root VerkleNode, keys [][]byte, resolver NodeResolverFn) (*Proof, []*Point, []byte, []*Fr, error) {
+func getProofElementsFromTree(preroot, postroot VerkleNode, keys [][]byte, resolver NodeResolverFn) (*ProofElements, []byte, [][]byte, [][]byte, []*Point, [][]Fr, []*Fr, []byte, error) {
 	// go-ipa won't accept no key as an input, catch this corner case
 	// and return an empty result.
 	if len(keys) == 0 {
-		return nil, nil, nil, nil, errors.New("no key provided for proof")
+		return nil, nil, nil, nil, nil, nil, nil, nil, errors.New("no key provided for proof")
 	}
 
-	tr := common.NewTranscript("vt")
-	root.Commit()
-
-	pe, es, poas, err := GetCommitmentsForMultiproof(root, keys, resolver)
+	pe, es, poas, err := GetCommitmentsForMultiproof(preroot, keys, resolver)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting pre-state proof data: %w", err)
+	}
+
+	// List of points and vectors and indices used to generate the IPA proof
+	// If only a pre-state root is available, it's a simple copy, but if both
+	// pre- and post-state root are present, it's merging the two lists.
+	var (
+		proof_cis []*Point
+		proof_fs  [][]Fr
+		proof_zs  []uint8
+		proof_ys  []*Fr
+		postvals  = make([][]byte, len(keys))
+	)
+	for i := range pe.Cis {
+		proof_cis = append(proof_cis, pe.Cis[i])
+		proof_fs = append(proof_fs, pe.Fis[i])
+		proof_ys = append(proof_ys, pe.Yis[i])
+		proof_zs = append(proof_zs, pe.Zis[i])
+	}
+
+	if postroot != nil {
```

Comment:
This block concatenates the post- and pre-state data together, to be used in the proof.